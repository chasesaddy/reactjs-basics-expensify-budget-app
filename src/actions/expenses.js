import uuid from 'uuid';
import db from '../firebase/firebase';

// 
//// Expenses
// 

// ADD_EXPENSE

const addExpense = ( expense ) => ({
  type: 'ADD_EXPENSE',
  expense
});

const startAddExpense = ( expenseData = {} ) => {
  return ( dispatch ) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return db.ref( 'expenses' ).push( expense ).then( ( ref ) => { 
      dispatch( addExpense( { 
        id: ref.key,
        ...expense
      } ) );
    } );
  };
};

// REMOVE_EXPENSE

const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

const startRemoveExpense = ( { id } = {} ) => {
  return ( dispatch ) => {
    return db.ref( `expenses/${ id }` ).remove().then( ( ref ) => { 
      dispatch( removeExpense( { id } ) );
    } );
  };
};

// EDIT_EXPENSE

const editExpense = ( id, updates ) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES
const setExpenses = ( expenses ) => ( { 
  type: 'SET_EXPENSES',
  expenses
} );

const startSetExpenses = () => {
  return ( dispatch ) => {
    return db.ref( 'expenses' ).once( 'value' ).then( ( snapshot ) => { 
      const allExpenses = [];
      snapshot.forEach( ( solo ) => { 
        allExpenses.push( {
          id: solo.key,
          ...solo.val()
        } );
      } );      
      dispatch( setExpenses( allExpenses ) );
    } );
  };
};

export { addExpense, startAddExpense, removeExpense, startRemoveExpense, editExpense, setExpenses, startSetExpenses };
