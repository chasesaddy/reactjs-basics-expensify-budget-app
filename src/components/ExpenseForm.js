import React from 'react';

import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends React.Component {
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    calFocused: false,
    errorMsg: ''
  };

  onDescriptionChange = ( e ) => {
    const description = e.target.value;
    this.setState( () => ( { description } ) );
  };

  onNoteChange = ( e ) => {
    const note = e.target.value;
    this.setState( () => ( { note } ) );
  };

  onAmountChange = ( e ) => {
    const amount = e.target.value;
    if ( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ) {
      this.setState( () => ( { amount } ) );
    }
  };

  onDateChange = ( createdAt ) => {
    if ( createdAt ) {
      this.setState( () => ( { createdAt } ) );
    }
  };

  onFocusChange  = ( { focused } ) => {
    this.setState( () => ( { calFocused: focused } ) );
  };

  onSubmit = ( e ) => {
    e.preventDefault();

    if ( !this.state.description || !this.state.amount ) {
      this.setState( () => ( { error: 'Need description and amount, please.' } ) );
    } else {
      this.setState( () => ( { error: '' } ) );
      console.log( 'Submitted' );
    }
  }

  render() {
    return (
      <div>
        
          { this.state.error && 
          <div className="errorMsg">{ this.state.error }</div>
          }        

        <form onSubmit={ this.onSubmit }>

          <input 
            type="text"
            placeholder="Description"
            autoFocus
            value={ this.state.description }
            onChange={ this.onDescriptionChange }
          />

          <input 
            type="text"
            placeholder="Amount"
            value={ this.state.amount }
            onChange={ this.onAmountChange }
          />

          <SingleDatePicker 
            date={ this.state.createdAt }
            onDateChange={ this.onDateChange }
            focused={ this.state.calFocused }
            onFocusChange={ this.onFocusChange }
            numberOfMonths={ 1 }
            isOutsideRange={ () => { false } }
          />

          <textarea
            placeholder="Add a note for your expense (optional)"
            value={ this.state.note }
            onChange={ this.onNoteChange }
          >
          </textarea>

          <button>
            Add Expense
          </button>

        </form>

      </div>
    )
  };
}

export default ExpenseForm;
