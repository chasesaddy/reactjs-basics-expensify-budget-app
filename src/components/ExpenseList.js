import React from 'react';
import { connect } from 'react-redux';

const ExpenseList = ( props ) => (
  <div>
    <h1>
      Expense List
    </h1>

    <div>
      <ul>
        { props.filters.text }
        { props.expenses.map( ( solo ) => <li key={ solo.id }>{ solo.description }</li> ) }
      </ul>
    </div>
  </div>
);

const mapStateToProps = ( state ) => {
  return {
    expenses: state.expenses
    , filters: state.filters
  };
};

export default connect( mapStateToProps )( ExpenseList );
