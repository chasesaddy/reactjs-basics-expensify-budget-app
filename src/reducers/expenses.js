// Expenses Reducer

const expensesReducerDefaultState = [];

const expensesReducer = ( state = expensesReducerDefaultState, action ) => {
  switch ( action.type ) {
  case 'ADD_EXPENSE':
    return [
      ...state, action.expense
    ];
  case 'REMOVE_EXPENSE': 
    console.log( 'here' );
    console.log( action.id );
    return state.filter( ( oneExpense ) => oneExpense.id !== action.id );
    // Could have done destructuring. Didn't think of that.
    // return state.filter( ( { id } ) => id !== action.id );
  case 'EDIT_EXPENSE': 
    return state.map( ( oneExpense ) => {
      if ( oneExpense.id === action.id ) {
        return {
          ...oneExpense,
          ...action.updates
        };
      } else {
        return oneExpense;
      }
    });
  default:
    // will print filterReducer action.type
    console.log( action.type );
    return state;
  }
};

export default expensesReducer;
