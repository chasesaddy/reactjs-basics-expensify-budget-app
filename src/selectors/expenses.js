import moment from 'moment';

const getVisibleExpenses = ( expenses, { text, sortBy, startDate, endDate } ) => {
  return expenses.filter( ( solo ) => {
    
    const createdAtMoment = moment( solo.createdAt )

    const startDateMatched = startDate ? moment( startDate ).isSameOrBefore( createdAtMoment, 'day' ) : true;
    const endDateMatched = startDate ? moment( endDate ).isSameOrAfter( createdAtMoment, 'day' ) : true;
    const textMatched = textMatch( solo.description, text );

    return startDateMatched && endDateMatched && textMatched;
  }).sort( ( a, b ) => {
    if ( sortBy == 'date' ) {
      return b.createdAt > a.createdAt ? 1 : -1;
    } else if ( sortBy == 'amount' ) {
      return b.amount > a.amount ? 1 : -1;
    }
  });
};

const startDateMatch = ( objCreatedDate, startDate ) => {
  return ( typeof startDate !== 'number' || objCreatedDate >= startDate );
};

const endDateMatch = ( objCreatedDate, endDate ) => {
  return ( typeof endDate !== 'number' || objCreatedDate <= endDate );
};

const textMatch = ( objText, text ) => {
  return ( typeof text !== 'string' || objText.toLowerCase().includes( text.toLowerCase() ) );
};

export default getVisibleExpenses;
