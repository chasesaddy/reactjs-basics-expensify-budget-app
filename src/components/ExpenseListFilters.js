import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calFocused: null,
    focusedInput: null
  };

  onDatesChange = ( { startDate, endDate } ) => {
    this.props.setStartDate( startDate );
    this.props.setEndDate( endDate );

    this.setState( {
      focusedInput: startDate && !endDate ? 'startDate' : 'endDate'
    } );
  };

  onFocusChange = ( calFocused ) => {
    this.setState( () => ( { calFocused } ) );
  };

  onTextChange = ( e ) => {
    this.props.setTextFilter( e.target.value );
  }

  onSortChange = ( e ) => {
    if ( e.target.value === 'date' ) {
      this.props.sortByDate();
    } else if ( e.target.value === 'amount' ) {
      this.props.sortByAmount();
    }
  }


  render() {
    return (
      <div>
        <input 
          type="text" 
          value={ this.props.filters.text } 
          onChange={ this.onTextChange }
        />

        <select 
          value={ this.props.filters.sortBy }
          onChange={ this.onSortChange }
        >
          <option value="date">
            Date
          </option>
          <option value="amount">
            Amount
          </option>
        </select>
        <DateRangePicker 
          startDate={ this.props.filters.startDate }
          endDate={ this.props.filters.endDate }
          startDateId="start"
          endDateId="end"
          onDatesChange={ this.onDatesChange }
          focusedInput={ this.state.focusedInput }
          onFocusChange={ this.onFocusChange }
          showClearDates={ true }
          numberOfMonths={ 1 }
          isOutsideRange={ () => false }
        />
      </div>
    );
  }
};

const mapStateToProps = ( state ) => ( {
  filters: state.filters
} );

const MapDispatchToProps = ( dispatch ) => ( { 
  setStartDate: ( startDate ) => dispatch( setStartDate( startDate ) ),
  setEndDate: ( endDate ) => dispatch( setEndDate( endDate ) ),
  setTextFilter: ( text ) => dispatch( setTextFilter( text ) ),
  sortByDate: () => dispatch( sortByDate() ),
  sortByAmount: () => dispatch( sortByAmount() ),
} );

export default connect( mapStateToProps, MapDispatchToProps )( ExpenseListFilters );
