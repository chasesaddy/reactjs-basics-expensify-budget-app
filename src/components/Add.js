import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = ( oneExpense ) => {    
    this.props.onSubmit( oneExpense );
    this.props.history.push( '/' );
  };

  render() {
    return (
      <div>
        <h1>
          Add Expense
        </h1>

        <ExpenseForm 
          onSubmit={ this.onSubmit }
        />
      </div>
    );
  };
}

const mapsDispatchToProps = ( dispatch ) => ( {
  onSubmit: ( expense ) => dispatch( addExpense ( expense ) )
} );

export default connect( undefined, mapsDispatchToProps )( AddExpensePage );
