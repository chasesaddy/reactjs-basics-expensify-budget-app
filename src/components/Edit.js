import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = ( oneExpense ) => {
    this.props.editExpense( this.props.expense.id, oneExpense );
    this.props.history.push( '/' );
  };

  onClick = () => { 
    const theId = this.props.expense.id;
    
    this.props.removeExpense( theId );
    this.props.history.push( '/' );
  };


  render() {
    return (
      <div>
        <ExpenseForm 
          expense={ this.props.expense }
          onSubmit={ this.onSubmit }
        />
        <button onClick={ this.onClick }>
          Remove
        </button>
      </div>
    );
  };
}

const mapStateToProps = ( state, props ) => {
  return {
    expense: state.expenses.find( ( oneExpense ) => oneExpense.id === props.match.params.id )
  };
};

const mapDispatchToProps = ( dispatch ) => ( { 
  editExpense: ( theId, oneExpense ) => dispatch( editExpense( theId, oneExpense ) ),  
  removeExpense: ( theId ) => dispatch( removeExpense( { id: theId } ) )
} );

export default connect( mapStateToProps )( EditExpensePage );
