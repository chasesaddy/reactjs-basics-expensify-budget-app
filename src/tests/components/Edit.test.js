import React from 'react';
import { shallow } from 'enzyme';

import { EditExpensePage } from '../../components/Edit';

import expenses from '../fixtures/expenses';

let editExpense, removeExpense, history, wrapper, example;
beforeEach( () => { 
  editExpense = jest.fn();
  removeExpense = jest.fn();
  
  history = { push: jest.fn() };

  example = expenses[ 0 ];

  wrapper = shallow( <EditExpensePage 
    expense={ example } 
    history={ history } 
    editExpense={ editExpense } 
    removeExpense={ removeExpense } 
  /> )
} );

test( 'should render EditExpensePage correctly', () => { 
  expect( wrapper ).toMatchSnapshot();
} );

test( 'should handle onSubmit which is editExpense', () => {   
  wrapper.find( 'ExpenseForm' ).prop( 'onSubmit' )( example );

  expect( history.push ).toHaveBeenLastCalledWith( '/' );
  expect( editExpense ).toHaveBeenLastCalledWith( example.id, example );
} );

test( 'should handle onClick which is removeExpense', () => { 
  wrapper.find( 'button' ).prop( 'onClick' )();

  expect( history.push ).toHaveBeenLastCalledWith( '/' );
  expect( removeExpense ).toHaveBeenLastCalledWith( example.id );
} );
