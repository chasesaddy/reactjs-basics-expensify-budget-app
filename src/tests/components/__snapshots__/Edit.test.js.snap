// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 350,
        "createdAt": -345600000,
        "description": "Gumming",
        "id": "1",
        "note": "Shiz just keeps on...",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
